examples/examples.rs

```diff
@@ -82,37 +80,26 @@ fn examples() -> Result<()> {
     // This API handles variadics using Heterogeneous Lists.  This is one way to
     // call a function with multiple parameters:
 
-    print.call::<_, ()>(
-        hlist!["hello", "again", "from", "rust"],
-    )?;
+    print.call::<_, ()>(("hello", "again", "from", "rust"))?;
 
     // You can bind rust functions to lua as well
 
-    let check_equal = lua.create_function(|lua, args| {
-        // Functions wrapped in lua receive their arguments packed together as
-        // MultiValue.  The first thing that most wrapped functions will do
-        // is "unpack" this MultiValue into its parts.  Due to lifetime type
-        // signature limitations, this cannot be done automatically from the
-        // function signature, but this will be fixed with ATCs.  Notice the use
-        // of the hlist macros again.
-        let hlist_pat![list1, list2] = lua.unpack::<HList![Vec<String>, Vec<String>]>(args)?;
-
+    let check_equal = lua.create_function(|_, (list1, list2): (Vec<String>, Vec<String>)| {
```

Comment:
Some thoughts and notes (you probably already know this, I'm just writing it down to get a better understanding of the issues at hand):

* In part, this is caused because `Lua` is essentially a heterogeneous collection. `Vec<T>` does not have this problem because its type always expresses the borrow if there is one, because it must contain the lifetime of the borrow.
* It might be possible to express the "shortest borrowed lifetime" as a lifetime parameter of `Lua`, which should solve this.
* My experiments indicate that a closure borrowing the outside `lua` will not compile. In particular, [this](https://play.rust-lang.org/?gist=d6474ac18c718bdf037e0f979c3d7c5e&version=stable) works, but not [this](https://play.rust-lang.org/?gist=3f15d51cf9573e14e446f7805163f4e0&version=stable), and I don't see a way to make the second version work. Therefore, the parameter needs to be kept.
* Compiler errors are extremely bad when the lifetimes are wrong. This does not make it pleasant to fix.

We should also use [compiletest](https://github.com/laumann/compiletest-rs) to make sure this doesn't regress to allowing bad programs to compile (there's another commented-out test somewhere that could be converted, too). Sadly, I haven't found that crate to be very pleasant to use, but it's all we've got for now.